### PR TITLE
apiserver: Use WaitAdvance in listenerSuite.TestPause

### PR DESCRIPTION
**Pull request by babbageclunk**
## Description of change

The test passes fine locally but fails often in CI, and when run under
stress it fails after 1 or 2 runs. It looks like the clock advance is
happening before the `throttlingListener` manages to wait for it. Change
the `Advance` calls to `WaitAdvance` so we know the listener is waiting when
we advance. The `time.After` threshold needs to be created before the
`WaitAdvance` call so that the done channel doesn't appear to be ready
early in the case that we needed to wait for the accept goroutine to be
waiting. (Picking the max waits on the `WaitAdvance` calls was fiddly
because `WaitAdvance` divides the max wait time into 10 chunks.)

## QA steps

The test passes normally and for more than 20 runs under stress.

### CODE REVIEW

File: apiserver/listener_test.go [hunk 1]
```diff
@@ -157,14 +158,17 @@ func (s *listenerSuite) TestPause(c *gc.C) {
...
-	case <-time.After(50 * time.Millisecond):
+	case <-minPauseTimeGone:
 	}
 	c.Assert(s.listener.count, gc.Equals, 0)
-	s.clock.Advance(time.Millisecond)
+	err = s.clock.WaitAdvance(10*time.Millisecond, time.Millisecond, 1)
+	c.Assert(err, jc.ErrorIsNil)
 	select {
 	case <-done:
 	case <-time.After(10 * time.Second):
```
Comment (axw) [inline apiserver/listener_test.go]: use coretesting.LongWait here? i.e. the amount of time to when something is expected to happen

File: apiserver/listener_test.go [hunk 2]
```diff
@@ -157,14 +158,17 @@ func (s *listenerSuite) TestPause(c *gc.C) {
 	}()
 
 	start <- true
-	s.clock.Advance((l.minPause/time.Millisecond - 1) * time.Millisecond)
+	minPauseTimeGone := time.After(50 * time.Millisecond)
+	err := s.clock.WaitAdvance(time.Millisecond, (l.minPause/time.Millisecond-1)*time.Millisecond, 1)
```
Comment (axw) [inline apiserver/listener_test.go]: use coretesting.ShortWait for the 2nd arg here? i.e. the amount of time to when something is expected *not* to happen

File: apiserver/listener_test.go [hunk 3]
```diff
@@ -157,14 +158,17 @@ func (s *listenerSuite) TestPause(c *gc.C) {
...
+	c.Assert(err, jc.ErrorIsNil)
 	select {
 	case <-done:
 		c.Fatal("pause returned too soon")
-	case <-time.After(50 * time.Millisecond):
+	case <-minPauseTimeGone:
```
Comment (axw) [inline apiserver/listener_test.go]: I think this should be coretesting.ShortWait too

File: apiserver/listener_test.go [hunk 4]
```diff
@@ -157,14 +158,17 @@ func (s *listenerSuite) TestPause(c *gc.C) {
 	}()
 
 	start <- true
```
Comment (babbageclunk) [inline apiserver/listener_test.go]: Good call - removed.

**Comment by babbageclunk** [comment]
$$merge$$

**Comment by babbageclunk** [comment]
Merge job died. :(

$$merge$$

**Comment by babbageclunk** [comment]
$$merge$$

**Pull request merged**